### DISCUSSION TITLE
While dualbooting Fedora and Windows. How to boot the windows partition from cockpit-machines?

### DISCUSSION THREAD

**Discussion by thepragmaticmero** [cockpit-project/cockpit-machines#1828] [Q&A]
My struggle:
I have a Surface tablet (so... no "just add one drive") with the following configuration:
```
nvme0n1p1    100MiB    windows boot manager
nvme0n1p2    16MiB     window recovery something
nvme0n1p3    250GiB    windows root partition
nvme0n1p4    200MiB    /boot/efi
nvme0n1p5    500MiB    /boot
nvme0n1p6    280GiB    / (linux root partition)
```
The windows partitions are currently unmounted. 
I've added them until I got a working VM in cockpit-machines. I added supergrub2.iso for debugging purposes and chainloading.

![Screenshot From 2024-09-20 16-12-21](https://github.com/user-attachments/assets/3aadff8e-cfe9-44e1-82b4-fb8a69152f84)
---
This is my [virsh output](https://gist.github.com/thepragmaticmero/da1f9b09033f52782a8775bd201bf296).
`sudo virsh edit ps_windowstest`

---
It does recognize supergrub2 which makes me japi japi japi, but I doesn't recognize the windows partition. I have no idea why.
![image](https://github.com/user-attachments/assets/c16a51aa-c191-4f7b-874c-a80493fd7e78)

Version: `cockpit-machines-319-1` `cockpit-system-324.1`
OS: Fedora Silverblue 40

Help please, I was trying to do this since 2022 and still have no idea what I'm doing wrong Q-Q.

Gatito blade for kitty purposes
![image](https://github.com/user-attachments/assets/1ad63220-8fe6-4812-8657-b3567a1a64bc)

**Comment by garrett** [comment]
I'm not sure exactly what you're trying to do.

I guess you have Windows on the disk and want to run it in a VM on Fedora within Cockpit-Machines? And I'm guessing once you have it working, you'd start the VM and use an RDP client, I guess to run Windows applications? (Especially since almost all games run on Linux these days thanks to Wine+Proton, and most types of apps are available in one form or another, so I'm assuming you're trying to access a specific application from Windows on Linux in a VM.)

I don't really have experience with this kind of task, but I'd suggest checking to see if your VM is set up for EFI or legacy BIOS. You probably have Windows installed using EFI (this is a very reasonable expectation in 2024 :wink:) and you would definitely need to have your VM use EFI for booting as well, not in legacy BIOS mode. That'd cause EFI partitions to not be seen, which lines up with what you're experiencing.

In Cockpit, you can check what your VM is by visiting its details page in the "Virtual machines" page and looking at "Overview" card, on the "Hypervisor" section. It should show you whether it is BIOS or EFI. Here's an example where it's BIOS:

![image](https://github.com/user-attachments/assets/9de2bcff-2589-495a-88fe-d6bdacad6a27)

If this is the issue you're hitting, it's rather involved to convert a machine from legacy BIOS to EFI, unfortunately. It can be done, but it's a bit convoluted and most people make a new VM with the correct settings instead.

You might be able to modify a shut down VM using virt-manager, installed with `rpm-ostree install --apply-live virt-manager` (the apply live, which also has a shortcut of `-A`, means to try to install an app on your system without rebooting... which usually works when changes aren't pending, otherwise it'll give you an error after installing and still require a reboot to activate it). But even with virt-manager, it might not be straightforward. You might still need to set up special configuration by hand editing XML inside of virsh.

There's a discussion @ https://unix.stackexchange.com/questions/612813/virt-manager-change-firmware-after-installation which might be useful. (Basically, it's probably easiest to create a new VM with UEFI instead of legacy BIOS, if that is the problem you're having.)

**Reply by thepragmaticmero** [reply]
I'm using EFI on windows and EFI on the VM. Still doesn't recognize. Note: It is **not** a conversion, I've created it with UEFI from scratch.
![image](https://github.com/user-attachments/assets/ac6b6733-bca9-4657-bcc2-1600f6427ff5)
I already modified the XML using virsh, it was fun.
My baremetal UEFI bios recognized Windows tho.
What disks should I add to the boot menu on _Boot order_?? I added the windows boot mgr and windows recovery partition and no boot. Then I added supergrub to see if I could chainboot. No UEFI entries. Weird huh. 

Replicating should give you more context. But asking to install another windows partition (and installing it without windows trying to merge grub and windows boot mgr) sounds like a lot to ask lol. 

> I'm not sure exactly what you're trying to do.
> I guess you have Windows on the disk and want to run it in a VM on Fedora within Cockpit-Machines?

Yep, that's what I'm trying to do. Nailed it. Reasoning: Save 30GB of windows install + not rebooting everytime I want to make small changes to windows (I'm setting up a dev env for targeting Win, Linux and Android, I need the testing.)

Also, thanks to the tip about `--apply-live` I didn't knew you could do that.

**Reply by garrett** [reply]
You should probably be aware of some issues with Windows licensing...

Windows stores your hardware configuration and compares it to what you're running, so even if you are successful with this, it will either get upset about changed hardware when you're in the VM or when you've rebooted back into Windows on bare metal. Or, likely: Both.

The "fix" is to either a bunch of complicated steps, and if that doesn't work: call Microsoft to have them reset it for your key. Every time. They  keep track of which systems you have that specific key on too, and probably won't let you do this so often. And if you got your copy of Windows with the computer (called an "OEM" key), then it's not transferable to a new machine.

So, IMHO: It's a better idea to keep them separate, unless it's a 1-way conversion.

More info: 

- https://support.microsoft.com/en-us/windows/reactivating-windows-after-a-hardware-change-2c0e962a-f04c-145b-6ead-fb3fc72b6665
- https://www.tomshardware.com/how-to/transfer-windows-license-to-new-pc

You are able to run an unactivated copy of Windows without putting in a key, but it'll have a watermark in the bottom right and you can't change the background wallpaper or theme and a few other things, but it mainly fully works otherwise. You might consider this for the VM (especially if you're using RDP on a per-window basis or have an app maximized anyway). It'll eat more space to have two Windows installations (on bare metal and in a VM), but 30 GB is not _that_ much these days, when we often have 0.5 - 2 TB on main storage on most systems (not only laptops and desktops, but even phones, cameras, and game consoles).

In other words, it's a headache to maintain the licensing mess to try to reuse an installation of Windows already installed on the disk and in a VM. If you don't do a full one way migration to the VM (which might not be easy and may have side effects), it's better off to have two separate installations.

If you are running Windows on Linux on the same machine (instead of a server), the easiest ways of doing so would be to use GNOME Boxes (regardless of which desktop you're on; it'll work on KDE and others in addition to GNOME), virt-manager or [quickemu](https://github.com/quickemu-project/quickemu).

(I use Boxes for all sorts of things, even on my development machine where I also have Cockpit Machines, and I use quickemu for testing some things too. I do have virt-manager installed and use it sometimes, but generally stick to Boxes when I can for easy desktop VMs or quickemu as a fallback.)

**Reply by thepragmaticmero** [reply]
I'll see if I can replicate it with a baremetal Arch Linux install as target. Still looks pretty sus to me ngl. I'll update this thread later then.
